streams/src/main/java/org/apache/kafka/streams/kstream/JoinWindows.java

```diff
@@ -246,7 +246,7 @@ public long gracePeriodMs() {
      * @param durationMs the window retention time in milliseconds
      * @return itself
      * @throws IllegalArgumentException if {@code durationMs} is smaller than the window size
-     * @deprecated since 2.1. Use {@link Materialized#withRetention(Duration)} instead.
+     * @deprecated since 2.1. Use {@link JoinWindows#grace(Duration)} instead.
```

Comment:
But grace and retention are two different things.

In fact, I just a another conversation about this issue, and it seem we need to fix this by allowing people to specify  a retention time, ie, we should add a `Joined#withRetention()` IMHO.

Reply:
Thanks, @mjsax . I think the ticket author is the same person you were talking with. He (later on) included a snippet of a chat with you.

I think join windows are a little confusing because we tried to share the logic of windowed aggregations, but it's actually different.

For windowed aggregation, grace period defines how long after the window ends it will accept late-arriving records and update the results. The retention time defines how long we keep the state of the window in storage. Clearly, the retention time must be at least big enough to support the updates that may happen during the grace period, but it could be much larger, to support Interactive Queries even after the window is closed to updates.

Join windows are a little bit different, because they are not queriable. Therefore, there is no reason to have any retention beyond the grace period. This is also why there's no `Materialized` parameter. The state for the join is purely bookeeping, not a "materialized view" in the data processing sense.

Reply:
The report included the additional issue, that the `gracePeriod()` and neither deprecated `until()` is forwarded as changelog configuration though. Can we include a fix for this on this ticket?

Reply:
If grace period isn't specified, then we should be forwarding the `until` aka `maintainMs` through. In his case, though, he was specifying grace period, so we were using it.

I think his bug report is partly based on a flawed understanding of what he should expect to see, but it's possible there's a real bug somewhere. I'm waiting to see what he says next.

Reply:
I agree that for JoinWindows it is not necessary to let users specify retention since the underlying stores would not be queryable at all.

And with `until` specified, the retention of the store, as well as the total "joinable" processing period, should be join-window + grace-period.